File: rub/goldenVillage.rb
# ----------

# GOLDEN VILLAGE SCRAPER

# - parses data from https://www.gv.com.sg/GVMovies
# - extracts the following
    # thumbnail
    # poster
    # caption
        # h5 ng-binding
        # caption-times ng-binding

# ----------

require 'selenium-webdriver'
require 'nokogiri'
require 'json'

def fetch_and_parse_html(url)
  driver = Selenium::WebDriver.for :chrome
  driver.navigate.to url

  previous_height = driver.execute_script('return document.body.scrollHeight')
  loop do
    driver.execute_script('window.scrollTo(0, document.body.scrollHeight)')
    sleep 3
    current_height = driver.execute_script('return document.body.scrollHeight')
    break if current_height == previous_height
    previous_height = current_height
  end

  content = driver.page_source
  driver.quit

  content
rescue => e
  puts "Error fetching HTML: #{e.message}"
  nil
end

def extract_films(url)
  html_text = fetch_and_parse_html(url)
  return unless html_text

  doc = Nokogiri::HTML(html_text)

  gv_films = doc.css('[id^="nowMovieThumb"]') 

  gv_films_object = []

  if gv_films.any?
    gv_films.each do |gv_film|
      poster_src = gv_film.at_css('.poster img')&.[]('src') || 'N/A'
      title = gv_film.at_css('.caption h5')
      title_text = title ? title.text.strip : 'N/A'
      title_text = title_text.gsub(/[+\^*]+$/, '')
      duration = gv_film.at_css('.caption-times')
      duration_text = duration ? duration.text.strip : 'N/A'

      gv_films_object << {
        title: title_text,
        tags: nil,
        description: nil,
        duration: duration_text,
        rating: nil,
        book_tickets_url: "https://www.gv.com.sg/GVMovies",
        poster_src: poster_src
      }
    end

    puts "Number of GV movies => #{gv_films_object.length}"
    gv_films_object
  else
    puts 'No GV films found'
    nil
  end
rescue => e
  puts "Error extracting films: #{e.message}"
  nil
end

# ----- EXECUTION CODE -----
    # For testing

films = extract_films("https://www.gv.com.sg/GVMovies")
if films
  puts JSON.pretty_generate(films)
else
  puts 'No movies data were found'
end

File: lib/goldenVillage.js
// ----------

// GOLDEN VILLAGE SCRAPER

// - parses data from https://www.gv.com.sg/GVMovies
// - extracts the following
    // thumbnail
    // poster
    // caption
        // h5 ng-binding
        // caption-times ng-binding

// - dependancy installation
    // npm install playwright
    //  npx playwright install

// ----------

const { chromium } = require('playwright');

async function fetchAndParseHTML(url) {
    try {
        const browser = await chromium.launch();
        const page = await browser.newPage();
        await page.goto(url, { waitUntil: 'networkidle' });
        let previousHeight;
        let currentHeight = await page.evaluate(() => document.body.scrollHeight);
        while (true) {
            previousHeight = currentHeight;
            await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
            await page.waitForTimeout(3000); // wait for content to load
            currentHeight = await page.evaluate(() => document.body.scrollHeight);
            if (currentHeight === previousHeight) break;
        }
        const content = await page.content();
        await browser.close();
        return content;
    } catch (error) {
        console.error('Error fetching HTML:', error);
    }
}

async function extractFilmsGv(url) {
    const htmlText = await fetchAndParseHTML(url);
    if (htmlText) {
        const { JSDOM } = require('jsdom');
        const dom = new JSDOM(htmlText);
        const doc = dom.window.document;

        // -- debugging -- 

        // console.log(doc.title);
        // console.log(doc.body.innerHTML)
        // const gvFilms = doc.querySelectorAll('#nowMovieThumb');

        const gvFilms = doc.querySelectorAll('[id^="nowMovieThumb"]'); // for elements with id that start with nowMovieThumb

        const gvFilmsObject = [];

        if (gvFilms.length) {
            gvFilms.forEach(gvFilm => {

                    // console.log(gvFilm.innerHTML);

                    const poster = gvFilm.querySelector('.poster img');
                    const posterSrc = poster ? poster.getAttribute('src') : 'N/A';
                    const title = gvFilm.querySelector('.caption h5');
                    let titleText = title ? title.textContent.trim() : 'N/A';
                    titleText = titleText.replace(/[+\^*]+$/, '');
                    const duration = gvFilm.querySelector('.caption-times');
                    const durationText = duration ? duration.textContent.trim() : 'N/A';

                    // console.log(`
                    //         -----
                    //         Title: ${titleText}
                    //         Duration: ${durationText}
                    //         Poster src: ${posterSrc}
                    //         -----
                    //     `);

                    gvFilmsObject.push({
                        title: titleText,
                        tags: "N/A",
                        description: "N/A",
                        duration: durationText,
                        rating: "N/A",
                        bookTicketsUrl: "https://www.gv.com.sg/GVMovies",
                        posterSrc: posterSrc
                    });

            });
            console.log(`number of gv movies => ${gvFilmsObject.length}`);
            return gvFilmsObject;
        } else {
            console.log('No gv films found');
        }
    } else {
        console.log('HTML content not found');
    }
};

module.exports = { extractFilmsGv };

// ----- EXECUTION CODE -----
    // - for testing
    // - actual usage to be called from /src/main.js

// extractFilmsGv("https://www.gv.com.sg/GVMovies").then((gvFilmsObject) => {
//     if (gvFilmsObject) {
//         console.log(JSON.stringify(gvFilmsObject, null, 2));
//     } else {
//         console.log('no movies data were found');
//     }
// });
